NeuroID/Utilities/NIDConfigService.swift

```diff
@@ -6,56 +6,95 @@ import Alamofire
-                     
-        let config_url = NIDConfigService.nidURL + NeuroID.clientKey! + ".json"
-        AF.request(config_url, method: .get).responseDecodable(of: ResponseData.self) { response in
+        
+        let config_url = NIDConfigService.NID_CONFIG_URL + NeuroID.clientKey! + ".json"
+        networkService.getRequest(
+            url: URL(string: config_url)!,
+            responseDecodableType: ConfigResponseData.self
+        ) { response in
             switch response.result {
             case .success(let responseData):
                 self.setCache(responseData)
-                NIDLog.d("Retrieved config log")
-                NIDConfigService.cacheSetWithRemote = true
-                completion(true)
+                NIDLog.d("Retrieved remote config")
+                
+                self.cacheSetWithRemote = true
+                self.cacheCreationTime = .init()
+                completion()
             case .failure(let error):
                 NIDLog.e("Failed to retrieve NID Config \(error)")
-                NIDConfigService.nidConfigCache = ResponseData()
-                completion(true)
+                self.configCache = ConfigResponseData()
+                self.cacheSetWithRemote = false
+                completion()
             }
         }
     }
     
-    func setCache(_ responseData: ResponseData) {
-        NIDConfigService.nidConfigCache = responseData
+    func setCache(_ newCache: ConfigResponseData) {
+        configCache = newCache
+    }
+    
+    func expiredCache() -> Bool {
+        if !cacheSetWithRemote {
+            return true
+        }
+        // 5 min is the default, can be updated later
+        let TTLTime = Calendar.current.date(byAdding: .minute, value: -5, to: Date())!
+        
+        return cacheCreationTime < TTLTime
+    }
+    
+    func retrieveOrRefreshCache(completion: @escaping () -> Void) {
+        if expiredCache() {
+            retrieveConfig {
+                completion()
+            }
+        } else {
+            completion()
+        }
+    }
+    
+    func updateConfigOptions(siteID: String? = nil, completion: @escaping () -> Void) {
+        // check cache time
+        retrieveOrRefreshCache {
+            // retrieve the site config for the site
+            
+            // if siteID == config.siteID - then use top level value
+            // if siteID == nil then assume parent site
+            if siteID == nil || siteID ?? "" == self.configCache.siteID ?? "noID" {
+                self.configCache.currentSampleRate = self.configCache.sampleRate ?? NIDConfigService.DEFAULT_SAMPLE_RATE
+                completion()
+                return
+            }
+            
+            // get linked site options and override config
+            let linkedSiteConfig = self.configCache.linkedSiteOptions?[siteID ?? ""]
+            if linkedSiteConfig == nil {
+                self.configCache.currentSampleRate = NIDConfigService.DEFAULT_SAMPLE_RATE
```

Comment:
Personally I would prefer a more immutable approach to the configCache. Ideally the configCache just holds sample rates by siteId/linkedSiteId, and does not have to do any logic to resolve "what the flow the user is currently in" 

Then in the `determineIsSessionSampled` function, we resolve the siteId to the available sites in the cache to find the sample rate.

This approach will **decouple** this logic from the simple cache, and push the business logic to where it belongs (`determineIsSessionSampled`)

Reply:
I thought about that approach however because linkedSites have the opportunity to have more options than just `sample_rate` I did not think `determineIsSessionSampled` was the right spot.

IMO the `determineIsSessionSampled`  function is just to use the config value and determine true/false. It's not to try and get an entire linkedSiteOptions and use them. This function however is built to update the config values based on the site/flow we are using.

I do agree we could make the configCache more immutable and maybe have a temporary holding config that represents the options for the linkedSite, however that seemed more appropriate when we expand the linkedSiteOptions than right now.

Reply:
`NidConfigService` operates as a caching layer for configs/linkedConfigs applicable to that site.

Multiple service can consume those configs for their own purposes, and there could be many different options within the configs. The service consuming the config should have the business logic of "what it needs to do"

Determining the sample rate is business logic that exist in the service that is designated to determine the sample rate, even if its required to pull all the sample rates of the linkedSiteIds.

I feel this is a very clear distinction, (think single responsibility principle for your services), making the code easier to read and less prone to bugs longer term.

Reply:
For example: if this service is async refreshing the cache in the background (this is an assumption but generally how a cache works) the service can refresh the cache _without_ calling this business logic.

Reply:
I understand what you are suggesting, I think we should discuss post standup/on zoom because I think this approach will result in more duplicate code